cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,263 @@
+  char quote_char = '"';
+
+  [[nodiscard]] inference_options_view view() const
+  {
+    return {cudf::detail::make_trie_view(trie_true),
+            cudf::detail::make_trie_view(trie_false),
+            cudf::detail::make_trie_view(trie_na),
+            quote_char};
+  }
+};
+
+/**
+ * @brief Returns true is the input character is a valid digit.
+ * Supports both decimal and hexadecimal digits (uppercase and lowercase).
+ *
+ * @param c Character to check
+ * @param is_hex Whether to check as a hexadecimal
+ *
+ * @return `true` if it is digit-like, `false` otherwise
+ */
+__device__ __inline__ bool is_digit(char const c, bool const is_hex = false)
+{
+  if (c >= '0' && c <= '9') return true;
+
+  if (is_hex) {
+    if (c >= 'A' && c <= 'F') return true;
+    if (c >= 'a' && c <= 'f') return true;
+  }
+
+  return false;
+}
+
+/**
+ * @brief Returns true if the counters indicate a potentially valid float.
+ * False positives are possible because positions are not taken into account.
+ * For example, field "e.123-" would match the pattern.
+ */
+__device__ __inline__ bool is_like_float(
+  long len, long digit_cnt, long decimal_cnt, long dash_cnt, long exponent_cnt)
+{
+  // Can't have more than one exponent and one decimal point
+  if (decimal_cnt > 1) return false;
+  if (exponent_cnt > 1) return false;
+  // Without the exponent or a decimal point, this is an integer, not a float
+  if (decimal_cnt == 0 && exponent_cnt == 0) return false;
+
+  // Can only have one '-' per component
+  if (dash_cnt > 1 + exponent_cnt) return false;
+
+  // If anything other than these characters is present, it's not a float
+  if (digit_cnt + decimal_cnt + dash_cnt + exponent_cnt != len) return false;
+
+  // Needs at least 1 digit, 2 if exponent is present
+  if (digit_cnt < 1 + exponent_cnt) return false;
+
+  return true;
+}
+
+template <typename ColumnStringIter>
+__global__ void detect_column_type_kernel(inference_options_view const options,
+                                          device_span<char const> const data,
+                                          ColumnStringIter column_strings_begin,
+                                          std::size_t const size,
+                                          cudf::io::column_type_histogram* column_info)
+{
+  auto idx = threadIdx.x + blockDim.x * blockIdx.x;
+
+  while (idx < size) {
+    auto const field_offset = thrust::get<0>(*(column_strings_begin + idx));
+    auto const field_len    = thrust::get<1>(*(column_strings_begin + idx));
+    auto const field_begin  = data.begin() + field_offset;
+    if (cudf::detail::serialized_trie_contains(
+          options.trie_na, {field_begin, static_cast<std::size_t>(field_len)})) {
+      atomicAdd(&column_info->null_count, 1);
+      continue;
+    }
+
+    // Handling strings
+    if (field_len == 0) continue;
+    if (*field_begin == options.quote_char && field_begin[field_len - 1] == options.quote_char) {
+      atomicAdd(&column_info->string_count, 1);
```

Comment:
I did a quick profile of the overall nested JSON parser and recognised that the type inference is taking a good bit of time compared to, e.g., type conversion. To some degree, I assume we're congesting on global memory `atomicAdd`s. As a "quick win", I tried aggregating the "histogram" locally in shared memory first, adding from the local histogram to the global histogram only at the end of the thread block's lifetime. That cut the duration in half. 
I assume that branch divergence is potentially another root cause, but reducing branch divergence will be a bit more involved.

Reply:
Good to know, which benchmark are you looking at? I can take that as a reference to test shared memory optimization.